src/Akka.Streams.Kafka/Stages/Consumers/Abstract/BaseSingleSourceLogic.cs

```diff
@@ -0,0 +1,159 @@
+        /// </summary>
+        protected abstract void ConfigureSubscription();
+
+        /// <summary>
+        /// This is called when stage downstream is finished
+        /// </summary>
+        protected abstract void PerformShutdown();
+
+        /// <summary>
+        /// Makes this logic task finished
+        /// </summary>
+        protected void OnShutdown()
+        {
+            _completion.SetResult(NotUsed.Instance);
+        }
+
+        /// <summary>
+        /// Configures manual subscription
+        /// </summary>
+        /// <param name="subscription"></param>
+        protected void ConfigureManualSubscription(IManualSubscription subscription)
+        {
+            switch (subscription)
+            {
+                case Assignment assignment:
+                    ConsumerActor.Tell(new KafkaConsumerActorMetadata.Internal.Assign(assignment.TopicPartitions), SourceActor.Ref);
+                    TopicPartitions = TopicPartitions.Union(assignment.TopicPartitions);
+                    break;
+                case AssignmentWithOffset assignmentWithOffset:
+                    ConsumerActor.Tell(new KafkaConsumerActorMetadata.Internal.AssignWithOffset(assignmentWithOffset.TopicPartitions), SourceActor.Ref);
+                    TopicPartitions = TopicPartitions.Union(assignmentWithOffset.TopicPartitions.Select(tp => tp.TopicPartition));
+                    break;
+            }
+        }
+
+        private void MessageHandling(Tuple<IActorRef, object> args)
+        {
+            switch (args.Item2)
+            {
+                case KafkaConsumerActorMetadata.Internal.Messages<K, V> msg:
+                    // might be more than one in flight when we assign/revoke tps
+                    if (msg.RequestId == _requestId)
+                        _requested = false;
+
+                    foreach (var consumerMessage in msg.MessagesList)
+                        _buffer.Enqueue(consumerMessage);
+                    
+                    Pump();
+                    
+                    break;
+                
+                case Status.Failure failure:
+                    FailStage(failure.Cause);
+                    break;
+                
+                case Terminated terminated:
+                    FailStage(new ConsumerFailed());
+                    break;
+            }
+        }
+
+        private void Pump()
+        {
+            if (IsAvailable(_shape.Outlet))
+            {
+                if (_buffer.TryDequeue(out var message))
+                {
+                    Push(_shape.Outlet, _messageBuilder.CreateMessage(message));
+                    Pump();
+                }
+                else if (!_requested && TopicPartitions.Any())
+                {
+                    RequestMessages();
+                }
+            }
+        }
+
+        protected void RequestMessages()
+        {
+            _requested = true;
```

Comment:
So the `_requested` field is mostly just for accounting purposes, to make sure we're not re-requesting data when the buffer is already filling up, right?

Reply:
Ditto with `_requestId` ?

Reply:
Yes. And `_requestId` is used to make sure that we will wait for the last response from consumer actor before will reset `_requested` flag and request more messages for empty buffer again.

Reply:
Got it, makes sense